### ISSUE TITLE
Issues with more than three entries

### ISSUE THREAD

**Issue by TDFKAOlli** [milaq/openwrt_knockd#2]
Hi,
I have problems with knockd on Lede 17.01.4. I have a longer config for knockd as I use it to control internet access for different stations in my lan. Here I notice that three configuration entries run fine but a fourth one doesn't see to work. I do have 10 configuration blocks in my config file.

Additionally I notices:
1) config is expected to be /etc/knockd.conf and not /etc/config/knockd. I guess this is standard path in knockd.
2) config installed in second packages seems to have a newer format which doesn't seem to be documented anywhere. At least I couldn't find it.

Could you check or is it a problem of the original knockd ?

Thanks,
TDFKAOlli

Example config below:
[options]
	logfile = /var/log/knockd.log
	interface = br-lan

[Block1]
	sequence    = 50213,51113,50432
	seq_timeout = 5
	command     = /etc/knockdscript.sh -I A B
	tcpflags    = syn

[Free1]
	sequence    = 50432,51113,50213
	seq_timeout = 5
	command     = /etc/knockdscript.sh -D A B
	tcpflags    = syn

[Block2]
	sequence    = 43653,32243,49433
	seq_timeout = 5
	command     = /etc/knockdscript.sh -I C D
	tcpflags    = syn

[Free2]
	sequence    = 49433,32243,43653
	seq_timeout = 5
	command     = /etc/knockdscript.sh -D C D
	tcpflags    = syn

[Block3]
	sequence    = 37598,27584,59256
	seq_timeout = 5
	command     = /etc/knockdscript.sh -I E F
	tcpflags    = syn

[Free3]
	sequence    = 59256,27584,37598
	seq_timeout = 5
	command     = /etc/knockdscript.sh -D E F
	tcpflags    = syn

[Block4]
	sequence    = 37897,27584,59526
	seq_timeout = 5
	command     = /etc/knockdscript.sh -I G H
	tcpflags    = syn

[Free4]
	sequence    = 59526,27584,37897
	seq_timeout = 5
	command     = /etc/knockdscript.sh -D G H
	tcpflags    = syn

[AllBlock]
	sequence    = 57763,25576,34345
	seq_timeout = 5
	command     = /etc/knockdscript.sh -I A B C D E F G H 
	tcpflags    = syn

[AllFree]
	sequence    = 34345,25576,57763
	seq_timeout = 5
	command     = /etc/knockdscript.sh -D A B C D E F G H 
	tcpflags    = syn

**Comment by milaq** [comment]
> config is expected to be /etc/knockd.conf and not /etc/config/knockd. I guess this is standard path in knockd

this package doesn't use the default upstream knockd config path.

`/etc/config/knockd` is the UCI compatible config file which gets translated into the knockd syntax to `/var/run/knockd.conf` which in turn gets loaded via https://github.com/milaq/openwrt_knockd/blob/master/files/knockd.init#L46 (instead of the upstream default - `/etc/knockd.conf`)

> config installed in second packages seems to have a newer format which doesn't seem to be documented anywhere. At least I couldn't find it.

take a look here: https://wiki.openwrt.org/doc/uci#file_syntax
it's the default config syntax for the whole system. everything that goes into `/etc/config/` should have that syntax in order to make it possible to manage your configuration via `uci`.

just issue a `uci show knockd` and it should return your knockd configuration; that way you can also verify that you don't get any parsing errors.
you can also change your knockd configuration via `uci set ...` or show other system config bits with `uci show`.

> Here I notice that three configuration entries run fine but a fourth one doesn't see to work.

can you please post the contents of `/var/run/knockd.conf` after starting the daemon, so we can make sure everything gets parsed correctly. also, as mentioned above, make sure `uci show knockd` doesn't comlain about parsing errors.

**Comment by TDFKAOlli** [comment]
O.k. I understand now the concept of the uci config files and the translation to the config file in /var/run/knockd.conf. Thanks for that info.
In the end if I translate the config into the uci format and let it convert. But anyhow this doesn't change anthing because I get not the same config file (appart from leading spaces and I'm not able to redirect the log-output). 
In the syslog I see:
Wed Jan  3 20:35:11 2018 user.notice knockd: waiting for child processes...
Wed Jan  3 20:35:11 2018 user.notice knockd: shutting down
Wed Jan  3 20:35:11 2018 user.notice knockd: waiting for child processes...
Wed Jan  3 20:35:11 2018 user.notice knockd: shutting down
Wed Jan  3 20:35:19 2018 user.notice knockd: waiting for child processes...
Wed Jan  3 20:35:19 2018 user.notice knockd: shutting down
Wed Jan  3 20:35:25 2018 user.notice knockd: waiting for child processes...
Wed Jan  3 20:35:25 2018 user.notice knockd: shutting down
Wed Jan  3 20:35:31 2018 user.notice knockd: waiting for child processes...
Wed Jan  3 20:35:31 2018 user.notice knockd: shutting down
Wed Jan  3 20:35:36 2018 user.notice knockd: waiting for child processes...
Wed Jan  3 20:35:36 2018 user.notice knockd: shutting down
Wed Jan  3 20:35:42 2018 user.notice knockd: waiting for child processes...
Wed Jan  3 20:35:42 2018 user.notice knockd: shutting down
Wed Jan  3 20:35:42 2018 daemon.info procd: Instance knockd::instance1 s in a crash loop 6 crashes, 1 seconds since last crash

Thats why I run it at the command line using the "-D" option. Here I see a problöem with pcap getting confused with the tcpflags parameter of the fourth entry. So still, three works, the fourth one not.

**Comment by TDFKAOlli** [comment]
Thinking about it again, I'm sure it is a problem in the original knockd code. Your repo ist just providing a build framework for lede and the conversion into a valid config file works.

https://github.com/jvinet/knock/ doesn't seem to be continued. Could you build from latest master? Unfortunately it doesn't have tags so to be able to build the previously used 0.5 version. I'll try to setup the Lede build environment to try my own build.

**Comment by TDFKAOlli** [comment]
BTW I think I found the issue. See here:
https://github.com/jvinet/knock/issues/54

**Comment by milaq** [comment]
alright, thanks for reporting :)

**Comment by milaq** [comment]
i'll reopen, until we have a patch ready

**Comment by TDFKAOlli** [comment]
Setup the build environment as mentioned in the README. Actually I had issues as I first had to build libpcap before I could do the steps for knockd. So I ran:
make  packages/libs/libpcap/configure
make  packages/libs/libpcap/compile
make  packages/libs/libpcap/install

Afterwards I did the steps for knockd and they seem to work fine. I just do not have an .ipk. There is even no bin/packages folder where I would expect it. I also couldn't find it elsewhere. I checked the Lede build system User guide but couldn't find any additional hint. Looks like the compile stage is not executed at all.
I cloned lede and set branch to latest version v17.01.4.
Do you know whats could be going wrong? I could start with a patch when I can build the current knockd.

**Comment by TDFKAOlli** [comment]
O.k. I finally found that I have to add the <M> flag to the knockd package in "make menuconfig" in Networks->Firewall. Without this set, the make didn't do anything for the package.
Took me a bit to find out, so I guess it would be nice to be added to the README.
I created a patch and tested it and now it works fine on my router and I can read in more that 3 configuration blocks. I also added a smaller patch I have seen on the other repository and I thought would be good to have it in as well.
I created a pull request.
